// @bacons/xcode/build/api/IDEWorkspaceChecks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IDEWorkspaceChecks = void 0;
/**
 * High-level API for IDEWorkspaceChecks.plist files.
 *
 * Introduced in Xcode 9.3, these files store the state of workspace checks
 * to prevent them from being recomputed each time the workspace is opened.
 *
 * Currently known keys:
 * - IDEDidComputeMac32BitWarning: Tracks whether the 32-bit macOS deprecation
 *   warning has been computed/shown. Setting to true suppresses the warning.
 */
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
const checks_1 = require("../workspace/checks");
/** Relative path from workspace to the checks plist */
const CHECKS_PATH = "xcshareddata/IDEWorkspaceChecks.plist";
/**
 * High-level class for working with IDEWorkspaceChecks.plist files.
 */
class IDEWorkspaceChecks {
    constructor(props, filePath) {
        this.props = props;
        this.filePath = filePath;
    }
    /**
     * Open an existing IDEWorkspaceChecks.plist from a workspace.
     *
     * @param workspacePath Path to the .xcworkspace directory
     * @returns The checks instance, or null if the file doesn't exist
     */
    static open(workspacePath) {
        const checksPath = path_1.default.join(workspacePath, CHECKS_PATH);
        if (!(0, fs_1.existsSync)(checksPath)) {
            return null;
        }
        const plistString = (0, fs_1.readFileSync)(checksPath, "utf-8");
        const props = (0, checks_1.parseChecks)(plistString);
        return new IDEWorkspaceChecks(props, checksPath);
    }
    /**
     * Open an existing IDEWorkspaceChecks.plist or create a new one.
     *
     * @param workspacePath Path to the .xcworkspace directory
     * @returns The checks instance (opened or newly created)
     */
    static openOrCreate(workspacePath) {
        const existing = IDEWorkspaceChecks.open(workspacePath);
        if (existing) {
            return existing;
        }
        const checksPath = path_1.default.join(workspacePath, CHECKS_PATH);
        return new IDEWorkspaceChecks({ IDEDidComputeMac32BitWarning: true }, checksPath);
    }
    /**
     * Create a new IDEWorkspaceChecks instance.
     *
     * @param options Optional initial props and file path
     */
    static create(options) {
        const defaultProps = {
            IDEDidComputeMac32BitWarning: true,
        };
        const props = { ...defaultProps, ...options?.props };
        return new IDEWorkspaceChecks(props, options?.filePath);
    }
    /**
     * Save the checks to disk.
     *
     * @param filePath Optional path to save to. If not provided, uses this.filePath.
     */
    save(filePath) {
        const targetPath = filePath ?? this.filePath;
        if (!targetPath) {
            throw new Error("No file path specified. Either provide a path or set this.filePath.");
        }
        // Ensure parent directory exists
        const dir = path_1.default.dirname(targetPath);
        if (!(0, fs_1.existsSync)(dir)) {
            (0, fs_1.mkdirSync)(dir, { recursive: true });
        }
        const plistString = (0, checks_1.buildChecks)(this.props);
        (0, fs_1.writeFileSync)(targetPath, plistString, "utf-8");
        this.filePath = targetPath;
    }
    /**
     * Save the checks to a workspace directory.
     *
     * @param workspacePath Path to the .xcworkspace directory
     */
    saveToWorkspace(workspacePath) {
        const checksPath = path_1.default.join(workspacePath, CHECKS_PATH);
        this.save(checksPath);
    }
    /**
     * Get the plist representation of the checks.
     */
    toPlist() {
        return (0, checks_1.buildChecks)(this.props);
    }
    /**
     * Get whether the Mac 32-bit warning has been computed/dismissed.
     */
    get mac32BitWarningComputed() {
        return this.props.IDEDidComputeMac32BitWarning ?? false;
    }
    /**
     * Set whether the Mac 32-bit warning has been computed/dismissed.
     */
    set mac32BitWarningComputed(value) {
        this.props.IDEDidComputeMac32BitWarning = value;
    }
    /**
     * Get a check value by key.
     *
     * @param key The check key
     * @returns The boolean value, or undefined if not set
     */
    getCheck(key) {
        return this.props[key];
    }
    /**
     * Set a check value.
     *
     * @param key The check key
     * @param value The boolean value
     */
    setCheck(key, value) {
        this.props[key] = value;
    }
    /**
     * Remove a check.
     *
     * @param key The check key
     * @returns true if the check was removed, false if it didn't exist
     */
    removeCheck(key) {
        if (key in this.props) {
            delete this.props[key];
            return true;
        }
        return false;
    }
}
exports.IDEWorkspaceChecks = IDEWorkspaceChecks;
