// @bacons/xcode/build/api/XCScheme.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createBuildableReference = exports.XCScheme = void 0;
/**
 * High-level API for Xcode scheme files (.xcscheme).
 *
 * Provides convenient methods for creating and manipulating schemes.
 */
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
const scheme = __importStar(require("../scheme"));
const DEFAULT_BUILD_FOR = {
    running: true,
    testing: true,
    profiling: true,
    archiving: true,
    analyzing: true,
};
/**
 * High-level class for working with Xcode scheme files.
 */
class XCScheme {
    constructor(name, props, filePath) {
        this.name = name;
        this.props = props;
        this.filePath = filePath;
    }
    /**
     * Load a scheme from a file.
     */
    static open(filePath) {
        const xml = (0, fs_1.readFileSync)(filePath, "utf-8");
        const props = scheme.parse(xml);
        const name = path_1.default.basename(filePath, ".xcscheme");
        return new XCScheme(name, props, filePath);
    }
    /**
     * Create a new empty scheme.
     */
    static create(name, props) {
        const defaultProps = {
            version: "1.7",
            buildAction: {
                parallelizeBuildables: true,
                buildImplicitDependencies: true,
                entries: [],
            },
            testAction: {
                buildConfiguration: "Debug",
                selectedDebuggerIdentifier: "Xcode.DebuggerFoundation.Debugger.LLDB",
                selectedLauncherIdentifier: "Xcode.DebuggerFoundation.Launcher.LLDB",
                shouldUseLaunchSchemeArgsEnv: true,
                testables: [],
            },
            launchAction: {
                buildConfiguration: "Debug",
                selectedDebuggerIdentifier: "Xcode.DebuggerFoundation.Debugger.LLDB",
                selectedLauncherIdentifier: "Xcode.DebuggerFoundation.Launcher.LLDB",
                launchStyle: "0",
                useCustomWorkingDirectory: false,
                ignoresPersistentStateOnLaunch: false,
                debugDocumentVersioning: true,
                debugServiceExtension: "internal",
                allowLocationSimulation: true,
            },
            profileAction: {
                buildConfiguration: "Release",
                shouldUseLaunchSchemeArgsEnv: true,
                savedToolIdentifier: "",
                useCustomWorkingDirectory: false,
                debugDocumentVersioning: true,
            },
            analyzeAction: {
                buildConfiguration: "Debug",
            },
            archiveAction: {
                buildConfiguration: "Release",
                revealArchiveInOrganizer: true,
            },
        };
        // Merge provided props with defaults
        const mergedProps = { ...defaultProps };
        if (props) {
            Object.assign(mergedProps, props);
        }
        return new XCScheme(name, mergedProps);
    }
    /**
     * Create a scheme for a target with sensible defaults.
     */
    static createForTarget(target, name, projectPath) {
        const schemeName = name ?? target.props.name;
        const containerPath = projectPath
            ? `container:${path_1.default.basename(projectPath)}`
            : `container:${path_1.default.basename(target.getXcodeProject().filePath, "/project.pbxproj")}.xcodeproj`;
        const buildableRef = createBuildableReference(target, containerPath);
        const xcscheme = XCScheme.create(schemeName);
        // Add build target
        xcscheme.addBuildTarget(buildableRef);
        // Set launch target
        xcscheme.setLaunchTarget(buildableRef);
        return xcscheme;
    }
    /**
     * Save the scheme to disk.
     *
     * @param filePath Optional path to save to. If not provided, uses this.filePath.
     */
    save(filePath) {
        const targetPath = filePath ?? this.filePath;
        if (!targetPath) {
            throw new Error("No file path specified. Either provide a path or set this.filePath.");
        }
        // Ensure directory exists
        const dir = path_1.default.dirname(targetPath);
        if (!(0, fs_1.existsSync)(dir)) {
            (0, fs_1.mkdirSync)(dir, { recursive: true });
        }
        const xml = scheme.build(this.props);
        (0, fs_1.writeFileSync)(targetPath, xml, "utf-8");
        this.filePath = targetPath;
    }
    /**
     * Get the XML representation of the scheme.
     */
    toXML() {
        return scheme.build(this.props);
    }
    /**
     * Get the buildable reference for a target UUID from this scheme.
     */
    getBuildableReference(targetUuid) {
        const entries = this.props.buildAction?.entries ?? [];
        for (const entry of entries) {
            if (entry.buildableReference?.blueprintIdentifier === targetUuid) {
                return entry.buildableReference;
            }
        }
        return null;
    }
    /**
     * Add a target to the build action.
     */
    addBuildTarget(ref, buildFor) {
        if (!this.props.buildAction) {
            this.props.buildAction = {
                parallelizeBuildables: true,
                buildImplicitDependencies: true,
                entries: [],
            };
        }
        if (!this.props.buildAction.entries) {
            this.props.buildAction.entries = [];
        }
        // Check if already added
        const existing = this.props.buildAction.entries.find((e) => e.buildableReference?.blueprintIdentifier === ref.blueprintIdentifier);
        if (existing) {
            return;
        }
        const config = { ...DEFAULT_BUILD_FOR, ...buildFor };
        const entry = {
            buildForRunning: config.running,
            buildForTesting: config.testing,
            buildForProfiling: config.profiling,
            buildForArchiving: config.archiving,
            buildForAnalyzing: config.analyzing,
            buildableReference: ref,
        };
        this.props.buildAction.entries.push(entry);
    }
    /**
     * Add a testable reference to the test action.
     */
    addTestTarget(ref) {
        if (!this.props.testAction) {
            this.props.testAction = {
                buildConfiguration: "Debug",
                testables: [],
            };
        }
        if (!this.props.testAction.testables) {
            this.props.testAction.testables = [];
        }
        // Check if already added
        const existing = this.props.testAction.testables.find((t) => t.buildableReference?.blueprintIdentifier === ref.blueprintIdentifier);
        if (existing) {
            return;
        }
        this.props.testAction.testables.push({
            skipped: false,
            buildableReference: ref,
        });
    }
    /**
     * Set the launch target for the scheme.
     */
    setLaunchTarget(ref) {
        if (!this.props.launchAction) {
            this.props.launchAction = {
                buildConfiguration: "Debug",
            };
        }
        this.props.launchAction.buildableProductRunnable = {
            runnableDebuggingMode: "0",
            buildableReference: ref,
        };
        // Also set for profile action
        if (!this.props.profileAction) {
            this.props.profileAction = {
                buildConfiguration: "Release",
            };
        }
        this.props.profileAction.buildableProductRunnable = {
            runnableDebuggingMode: "0",
            buildableReference: ref,
        };
    }
    /**
     * Set an environment variable for the launch action.
     */
    setLaunchEnvironmentVariable(key, value, isEnabled = true) {
        if (!this.props.launchAction) {
            this.props.launchAction = {
                buildConfiguration: "Debug",
            };
        }
        if (!this.props.launchAction.environmentVariables) {
            this.props.launchAction.environmentVariables = [];
        }
        // Update if exists
        const existing = this.props.launchAction.environmentVariables.find((v) => v.key === key);
        if (existing) {
            existing.value = value;
            existing.isEnabled = isEnabled;
            return;
        }
        this.props.launchAction.environmentVariables.push({
            key,
            value,
            isEnabled,
        });
    }
    /**
     * Add a command line argument to the launch action.
     */
    addLaunchArgument(argument, isEnabled = true) {
        if (!this.props.launchAction) {
            this.props.launchAction = {
                buildConfiguration: "Debug",
            };
        }
        if (!this.props.launchAction.commandLineArguments) {
            this.props.launchAction.commandLineArguments = [];
        }
        // Check if already added
        const existing = this.props.launchAction.commandLineArguments.find((a) => a.argument === argument);
        if (existing) {
            existing.isEnabled = isEnabled;
            return;
        }
        this.props.launchAction.commandLineArguments.push({
            argument,
            isEnabled,
        });
    }
}
exports.XCScheme = XCScheme;
/**
 * Helper to create a BuildableReference from a PBXNativeTarget.
 */
function createBuildableReference(target, referencedContainer) {
    const productName = target.props.productReference?.getDisplayName() ?? `${target.props.name}.app`;
    return {
        buildableIdentifier: "primary",
        blueprintIdentifier: target.uuid,
        buildableName: productName,
        blueprintName: target.props.name,
        referencedContainer,
    };
}
exports.createBuildableReference = createBuildableReference;
