// @bacons/xcode/build/api/XCSharedData.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.XCSharedData = void 0;
/**
 * High-level API for Xcode shared data directories (xcshareddata).
 *
 * Provides unified access to schemes, breakpoints, and workspace settings
 * stored in xcshareddata directories of .xcodeproj or .xcworkspace bundles.
 */
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
const breakpoints = __importStar(require("../breakpoints"));
const settings = __importStar(require("../settings"));
const scheme = __importStar(require("../scheme"));
const XCScheme_1 = require("./XCScheme");
/**
 * High-level class for working with Xcode shared data directories.
 *
 * Shared data includes:
 * - Schemes (xcschemes/*.xcscheme)
 * - Scheme management (xcschemes/xcschememanagement.plist)
 * - Breakpoints (xcdebugger/Breakpoints_v2.xcbkptlist)
 * - Workspace settings (WorkspaceSettings.xcsettings)
 */
class XCSharedData {
    constructor(filePath) {
        this._breakpointsLoaded = false;
        this._workspaceSettingsLoaded = false;
        this._schemeManagementLoaded = false;
        this.filePath = filePath;
    }
    /**
     * Open an xcshareddata directory.
     *
     * @param sharedDataPath Path to the xcshareddata directory
     */
    static open(sharedDataPath) {
        if (!(0, fs_1.existsSync)(sharedDataPath)) {
            throw new Error(`Shared data directory does not exist: ${sharedDataPath}`);
        }
        return new XCSharedData(sharedDataPath);
    }
    /**
     * Create a new XCSharedData instance.
     */
    static create() {
        return new XCSharedData();
    }
    // ============================================================================
    // Schemes
    // ============================================================================
    /**
     * Get the path to the xcschemes directory.
     */
    getSchemesDir() {
        if (!this.filePath)
            return undefined;
        return path_1.default.join(this.filePath, "xcschemes");
    }
    /**
     * Get all shared schemes.
     */
    getSchemes() {
        const schemesDir = this.getSchemesDir();
        if (!schemesDir || !(0, fs_1.existsSync)(schemesDir)) {
            return [];
        }
        const files = (0, fs_1.readdirSync)(schemesDir);
        return files
            .filter((f) => f.endsWith(".xcscheme"))
            .map((f) => XCScheme_1.XCScheme.open(path_1.default.join(schemesDir, f)));
    }
    /**
     * Get a scheme by name.
     */
    getScheme(name) {
        const schemesDir = this.getSchemesDir();
        if (!schemesDir)
            return null;
        const schemePath = path_1.default.join(schemesDir, `${name}.xcscheme`);
        if (!(0, fs_1.existsSync)(schemePath))
            return null;
        return XCScheme_1.XCScheme.open(schemePath);
    }
    /**
     * Save a scheme to the schemes directory.
     */
    saveScheme(xcscheme) {
        if (!this.filePath) {
            throw new Error("Cannot save scheme: no file path set for XCSharedData");
        }
        const schemesDir = path_1.default.join(this.filePath, "xcschemes");
        if (!(0, fs_1.existsSync)(schemesDir)) {
            (0, fs_1.mkdirSync)(schemesDir, { recursive: true });
        }
        const schemePath = path_1.default.join(schemesDir, `${xcscheme.name}.xcscheme`);
        xcscheme.save(schemePath);
    }
    /**
     * Get or load scheme management data.
     */
    get schemeManagement() {
        if (this._schemeManagementLoaded) {
            return this._schemeManagement;
        }
        this._schemeManagementLoaded = true;
        if (!this.filePath)
            return undefined;
        const managementPath = path_1.default.join(this.filePath, "xcschemes", "xcschememanagement.plist");
        if (!(0, fs_1.existsSync)(managementPath)) {
            return undefined;
        }
        const plistContent = (0, fs_1.readFileSync)(managementPath, "utf-8");
        this._schemeManagement = scheme.parseManagement(plistContent);
        return this._schemeManagement;
    }
    /**
     * Set scheme management data.
     */
    set schemeManagement(value) {
        this._schemeManagement = value;
        this._schemeManagementLoaded = true;
    }
    /**
     * Save scheme management to disk.
     */
    saveSchemeManagement() {
        if (!this.filePath) {
            throw new Error("Cannot save scheme management: no file path set for XCSharedData");
        }
        const schemesDir = path_1.default.join(this.filePath, "xcschemes");
        if (!(0, fs_1.existsSync)(schemesDir)) {
            (0, fs_1.mkdirSync)(schemesDir, { recursive: true });
        }
        const managementPath = path_1.default.join(schemesDir, "xcschememanagement.plist");
        if (this._schemeManagement) {
            const plistContent = scheme.buildManagement(this._schemeManagement);
            (0, fs_1.writeFileSync)(managementPath, plistContent, "utf-8");
        }
    }
    // ============================================================================
    // Breakpoints
    // ============================================================================
    /**
     * Get the path to the breakpoints file.
     */
    getBreakpointsPath() {
        if (!this.filePath)
            return undefined;
        return path_1.default.join(this.filePath, "xcdebugger", "Breakpoints_v2.xcbkptlist");
    }
    /**
     * Get or load breakpoints data.
     */
    get breakpoints() {
        if (this._breakpointsLoaded) {
            return this._breakpoints;
        }
        this._breakpointsLoaded = true;
        const breakpointsPath = this.getBreakpointsPath();
        if (!breakpointsPath || !(0, fs_1.existsSync)(breakpointsPath)) {
            return undefined;
        }
        const xml = (0, fs_1.readFileSync)(breakpointsPath, "utf-8");
        this._breakpoints = breakpoints.parse(xml);
        return this._breakpoints;
    }
    /**
     * Set breakpoints data.
     */
    set breakpoints(value) {
        this._breakpoints = value;
        this._breakpointsLoaded = true;
    }
    /**
     * Save breakpoints to disk.
     */
    saveBreakpoints() {
        if (!this.filePath) {
            throw new Error("Cannot save breakpoints: no file path set for XCSharedData");
        }
        const debuggerDir = path_1.default.join(this.filePath, "xcdebugger");
        if (!(0, fs_1.existsSync)(debuggerDir)) {
            (0, fs_1.mkdirSync)(debuggerDir, { recursive: true });
        }
        const breakpointsPath = path_1.default.join(debuggerDir, "Breakpoints_v2.xcbkptlist");
        if (this._breakpoints) {
            const xml = breakpoints.build(this._breakpoints);
            (0, fs_1.writeFileSync)(breakpointsPath, xml, "utf-8");
        }
    }
    // ============================================================================
    // Workspace Settings
    // ============================================================================
    /**
     * Get the path to the workspace settings file.
     */
    getWorkspaceSettingsPath() {
        if (!this.filePath)
            return undefined;
        return path_1.default.join(this.filePath, "WorkspaceSettings.xcsettings");
    }
    /**
     * Get or load workspace settings.
     */
    get workspaceSettings() {
        if (this._workspaceSettingsLoaded) {
            return this._workspaceSettings;
        }
        this._workspaceSettingsLoaded = true;
        const settingsPath = this.getWorkspaceSettingsPath();
        if (!settingsPath || !(0, fs_1.existsSync)(settingsPath)) {
            return undefined;
        }
        const plistContent = (0, fs_1.readFileSync)(settingsPath, "utf-8");
        this._workspaceSettings = settings.parse(plistContent);
        return this._workspaceSettings;
    }
    /**
     * Set workspace settings.
     */
    set workspaceSettings(value) {
        this._workspaceSettings = value;
        this._workspaceSettingsLoaded = true;
    }
    /**
     * Save workspace settings to disk.
     */
    saveWorkspaceSettings() {
        if (!this.filePath) {
            throw new Error("Cannot save workspace settings: no file path set for XCSharedData");
        }
        if (!(0, fs_1.existsSync)(this.filePath)) {
            (0, fs_1.mkdirSync)(this.filePath, { recursive: true });
        }
        const settingsPath = path_1.default.join(this.filePath, "WorkspaceSettings.xcsettings");
        if (this._workspaceSettings) {
            const plistContent = settings.build(this._workspaceSettings);
            (0, fs_1.writeFileSync)(settingsPath, plistContent, "utf-8");
        }
    }
    // ============================================================================
    // Save All
    // ============================================================================
    /**
     * Save all modified data to disk.
     *
     * @param dirPath Optional path to save to. If not provided, uses this.filePath.
     */
    save(dirPath) {
        const targetPath = dirPath ?? this.filePath;
        if (!targetPath) {
            throw new Error("No file path specified. Either provide a path or set this.filePath.");
        }
        this.filePath = targetPath;
        if (!(0, fs_1.existsSync)(targetPath)) {
            (0, fs_1.mkdirSync)(targetPath, { recursive: true });
        }
        if (this._breakpointsLoaded && this._breakpoints) {
            this.saveBreakpoints();
        }
        if (this._workspaceSettingsLoaded && this._workspaceSettings) {
            this.saveWorkspaceSettings();
        }
        if (this._schemeManagementLoaded && this._schemeManagement) {
            this.saveSchemeManagement();
        }
    }
}
exports.XCSharedData = XCSharedData;
