// @bacons/xcode/build/api/XCWorkspace.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.XCWorkspace = void 0;
/**
 * High-level API for Xcode workspace files (contents.xcworkspacedata).
 *
 * Provides convenient methods for creating and manipulating workspaces.
 */
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
const workspace = __importStar(require("../workspace"));
const IDEWorkspaceChecks_1 = require("./IDEWorkspaceChecks");
const XCSharedData_1 = require("./XCSharedData");
/**
 * High-level class for working with Xcode workspace files.
 */
class XCWorkspace {
    constructor(name, props, filePath) {
        this.name = name;
        this.props = props;
        this.filePath = filePath;
    }
    /**
     * Load a workspace from a .xcworkspace directory.
     *
     * @param workspacePath Path to the .xcworkspace directory
     */
    static open(workspacePath) {
        const contentsPath = path_1.default.join(workspacePath, "contents.xcworkspacedata");
        if (!(0, fs_1.existsSync)(contentsPath)) {
            throw new Error(`Invalid workspace: ${contentsPath} does not exist`);
        }
        const xml = (0, fs_1.readFileSync)(contentsPath, "utf-8");
        const props = workspace.parse(xml);
        const name = path_1.default.basename(workspacePath, ".xcworkspace");
        return new XCWorkspace(name, props, workspacePath);
    }
    /**
     * Create a new empty workspace.
     *
     * @param name The workspace name
     * @param props Optional initial workspace data
     */
    static create(name, props) {
        const defaultProps = {
            version: "1.0",
            fileRefs: [],
        };
        // Merge provided props with defaults
        const mergedProps = { ...defaultProps, ...props };
        return new XCWorkspace(name, mergedProps);
    }
    /**
     * Save the workspace to disk.
     *
     * @param workspacePath Optional path to save to. If not provided, uses this.filePath.
     */
    save(workspacePath) {
        const targetPath = workspacePath ?? this.filePath;
        if (!targetPath) {
            throw new Error("No file path specified. Either provide a path or set this.filePath.");
        }
        // Ensure .xcworkspace directory exists
        if (!(0, fs_1.existsSync)(targetPath)) {
            (0, fs_1.mkdirSync)(targetPath, { recursive: true });
        }
        const contentsPath = path_1.default.join(targetPath, "contents.xcworkspacedata");
        const xml = workspace.build(this.props);
        (0, fs_1.writeFileSync)(contentsPath, xml, "utf-8");
        this.filePath = targetPath;
    }
    /**
     * Get the XML representation of the workspace.
     */
    toXML() {
        return workspace.build(this.props);
    }
    /**
     * Add a project reference to the workspace.
     *
     * @param projectPath Path to the .xcodeproj relative to the workspace
     * @param locationType Location type prefix (default: "group")
     */
    addProject(projectPath, locationType = "group") {
        if (!this.props.fileRefs) {
            this.props.fileRefs = [];
        }
        const location = `${locationType}:${projectPath}`;
        // Check if already added
        const existing = this.props.fileRefs.find((ref) => ref.location === location);
        if (existing) {
            return;
        }
        this.props.fileRefs.push({ location });
    }
    /**
     * Remove a project reference from the workspace.
     *
     * @param projectPath Path to the .xcodeproj (with or without location prefix)
     * @returns true if the project was removed, false if not found
     */
    removeProject(projectPath) {
        if (!this.props.fileRefs) {
            return false;
        }
        const initialLength = this.props.fileRefs.length;
        // Match either the exact location or just the path portion
        this.props.fileRefs = this.props.fileRefs.filter((ref) => {
            const refPath = ref.location.includes(":")
                ? ref.location.split(":").slice(1).join(":")
                : ref.location;
            return refPath !== projectPath && ref.location !== projectPath;
        });
        return this.props.fileRefs.length < initialLength;
    }
    /**
     * Get all project paths in the workspace.
     *
     * @returns Array of project locations
     */
    getProjectPaths() {
        const paths = [];
        // Collect from top-level fileRefs
        if (this.props.fileRefs) {
            paths.push(...this.props.fileRefs.map((ref) => ref.location));
        }
        // Recursively collect from groups
        if (this.props.groups) {
            for (const group of this.props.groups) {
                paths.push(...this.collectPathsFromGroup(group));
            }
        }
        return paths;
    }
    /**
     * Add a group to the workspace.
     *
     * @param name The group name
     * @param location Optional location for the group
     * @returns The created group
     */
    addGroup(name, location) {
        if (!this.props.groups) {
            this.props.groups = [];
        }
        // Check if already exists
        const existing = this.props.groups.find((g) => g.name === name && g.location === location);
        if (existing) {
            return existing;
        }
        const group = {
            name,
            location,
            fileRefs: [],
        };
        this.props.groups.push(group);
        return group;
    }
    /**
     * Get a group by name.
     *
     * @param name The group name
     * @returns The group or undefined if not found
     */
    getGroup(name) {
        return this.props.groups?.find((g) => g.name === name);
    }
    /**
     * Remove a group from the workspace.
     *
     * @param name The group name
     * @returns true if the group was removed, false if not found
     */
    removeGroup(name) {
        if (!this.props.groups) {
            return false;
        }
        const initialLength = this.props.groups.length;
        this.props.groups = this.props.groups.filter((g) => g.name !== name);
        return this.props.groups.length < initialLength;
    }
    /**
     * Check if the workspace contains a project.
     *
     * @param projectPath Path to check (with or without location prefix)
     * @returns true if the project is in the workspace
     */
    hasProject(projectPath) {
        const paths = this.getProjectPaths();
        return paths.some((p) => {
            const pathPortion = p.includes(":") ? p.split(":").slice(1).join(":") : p;
            return pathPortion === projectPath || p === projectPath;
        });
    }
    /**
     * Get the IDEWorkspaceChecks for this workspace.
     *
     * @returns The checks instance, or null if not set
     */
    getWorkspaceChecks() {
        if (!this.filePath) {
            return null;
        }
        return IDEWorkspaceChecks_1.IDEWorkspaceChecks.open(this.filePath);
    }
    /**
     * Get or create the IDEWorkspaceChecks for this workspace.
     *
     * @returns The checks instance (opened or newly created)
     */
    getOrCreateWorkspaceChecks() {
        if (!this.filePath) {
            throw new Error("Workspace must be saved before accessing workspace checks.");
        }
        return IDEWorkspaceChecks_1.IDEWorkspaceChecks.openOrCreate(this.filePath);
    }
    /**
     * Check if this workspace has IDEWorkspaceChecks configured.
     *
     * @returns true if the checks plist exists
     */
    hasWorkspaceChecks() {
        if (!this.filePath) {
            return false;
        }
        return IDEWorkspaceChecks_1.IDEWorkspaceChecks.open(this.filePath) !== null;
    }
    /**
     * Mark the 32-bit warning as computed.
     *
     * This suppresses the macOS 32-bit deprecation warning dialog in Xcode
     * by setting IDEDidComputeMac32BitWarning to true.
     */
    setMac32BitWarningComputed() {
        const checks = this.getOrCreateWorkspaceChecks();
        checks.mac32BitWarningComputed = true;
        checks.save();
    }
    collectPathsFromGroup(group) {
        const paths = [];
        if (group.fileRefs) {
            paths.push(...group.fileRefs.map((ref) => ref.location));
        }
        if (group.groups) {
            for (const nestedGroup of group.groups) {
                paths.push(...this.collectPathsFromGroup(nestedGroup));
            }
        }
        return paths;
    }
    // ============================================================================
    // Shared Data Methods
    // ============================================================================
    /**
     * Get the path to the xcshareddata directory.
     */
    getSharedDataDir() {
        if (!this.filePath)
            return undefined;
        return path_1.default.join(this.filePath, "xcshareddata");
    }
    /**
     * Get the XCSharedData instance for this workspace.
     *
     * @returns XCSharedData for accessing schemes, breakpoints, and settings
     */
    getSharedData() {
        const sharedDataDir = this.getSharedDataDir();
        if (sharedDataDir && (0, fs_1.existsSync)(sharedDataDir)) {
            return XCSharedData_1.XCSharedData.open(sharedDataDir);
        }
        // Create a new instance with the path set
        const sharedData = XCSharedData_1.XCSharedData.create();
        if (sharedDataDir) {
            sharedData.filePath = sharedDataDir;
        }
        return sharedData;
    }
}
exports.XCWorkspace = XCWorkspace;
