// @bacons/xcode/build/breakpoints/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parse = void 0;
/**
 * Parser for Xcode breakpoint list files (Breakpoints_v2.xcbkptlist)
 *
 * Uses @xmldom/xmldom to parse XML into TypeScript objects.
 */
const xmldom_1 = require("@xmldom/xmldom");
/**
 * Parse a breakpoint list XML string into a typed XCBreakpointList object.
 */
function parse(xml) {
    const parser = new xmldom_1.DOMParser();
    const doc = parser.parseFromString(xml, "text/xml");
    const bucketEl = doc.documentElement;
    if (!bucketEl || bucketEl.tagName !== "Bucket") {
        throw new Error("Invalid breakpoint list file: root element must be <Bucket>");
    }
    return parseBucket(bucketEl);
}
exports.parse = parse;
function parseBucket(el) {
    const list = {};
    list.uuid = getAttr(el, "uuid");
    list.type = getAttr(el, "type");
    list.version = getAttr(el, "version");
    const breakpointsEl = getChildElement(el, "Breakpoints");
    if (breakpointsEl) {
        list.breakpoints = getChildElements(breakpointsEl, "BreakpointProxy").map(parseBreakpointProxy);
    }
    return list;
}
function parseBreakpointProxy(el) {
    const proxy = {
        breakpointExtensionID: getAttr(el, "BreakpointExtensionID") || "",
    };
    const contentEl = getChildElement(el, "BreakpointContent");
    if (contentEl) {
        proxy.breakpointContent = parseBreakpointContent(contentEl);
    }
    return proxy;
}
function parseBreakpointContent(el) {
    const content = {};
    content.uuid = getAttr(el, "uuid");
    content.shouldBeEnabled = getBoolAttr(el, "shouldBeEnabled");
    content.ignoreCount = getIntAttr(el, "ignoreCount");
    content.continueAfterRunningActions = getBoolAttr(el, "continueAfterRunningActions");
    content.filePath = getAttr(el, "filePath");
    content.startingColumnNumber = getAttr(el, "startingColumnNumber");
    content.endingColumnNumber = getAttr(el, "endingColumnNumber");
    content.startingLineNumber = getAttr(el, "startingLineNumber");
    content.endingLineNumber = getAttr(el, "endingLineNumber");
    content.landmarkName = getAttr(el, "landmarkName");
    content.landmarkType = getAttr(el, "landmarkType");
    content.condition = getAttr(el, "condition");
    content.scope = getAttr(el, "scope");
    content.symbolName = getAttr(el, "symbolName");
    content.moduleName = getAttr(el, "moduleName");
    content.exceptionType = getAttr(el, "exceptionType");
    content.stopOnStyle = getAttr(el, "stopOnStyle");
    const actionsEl = getChildElement(el, "Actions");
    if (actionsEl) {
        content.actions = getChildElements(actionsEl, "BreakpointActionProxy").map(parseBreakpointActionProxy);
    }
    const locationsEl = getChildElement(el, "Locations");
    if (locationsEl) {
        content.locations = getChildElements(locationsEl, "BreakpointLocationProxy").map(parseBreakpointLocation);
    }
    return content;
}
function parseBreakpointActionProxy(el) {
    const proxy = {
        actionExtensionID: getAttr(el, "ActionExtensionID") || "",
    };
    const contentEl = getChildElement(el, "ActionContent");
    if (contentEl) {
        proxy.actionContent = parseBreakpointActionContent(contentEl);
    }
    return proxy;
}
function parseBreakpointActionContent(el) {
    const content = {};
    content.consoleCommand = getAttr(el, "consoleCommand");
    content.message = getAttr(el, "message");
    content.conveyanceType = getAttr(el, "conveyanceType");
    content.shellCommand = getAttr(el, "shellCommand");
    content.shellArguments = getAttr(el, "shellArguments");
    content.waitUntilDone = getBoolAttr(el, "waitUntilDone");
    content.script = getAttr(el, "script");
    content.soundName = getAttr(el, "soundName");
    return content;
}
function parseBreakpointLocation(el) {
    // Handle the wrapper element (BreakpointLocationProxy)
    const locationContentEl = getChildElement(el, "BreakpointLocationContent");
    const targetEl = locationContentEl || el;
    const location = {};
    location.uuid = getAttr(targetEl, "uuid");
    location.shouldBeEnabled = getBoolAttr(targetEl, "shouldBeEnabled");
    location.ignoreCount = getIntAttr(targetEl, "ignoreCount");
    location.continueAfterRunningActions = getBoolAttr(targetEl, "continueAfterRunningActions");
    location.symbolName = getAttr(targetEl, "symbolName");
    location.moduleName = getAttr(targetEl, "moduleName");
    location.urlString = getAttr(targetEl, "urlString");
    location.startingLineNumber = getAttr(targetEl, "startingLineNumber");
    location.endingLineNumber = getAttr(targetEl, "endingLineNumber");
    location.startingColumnNumber = getAttr(targetEl, "startingColumnNumber");
    location.endingColumnNumber = getAttr(targetEl, "endingColumnNumber");
    return location;
}
// ============================================================================
// Helper functions
// ============================================================================
function getAttr(el, name) {
    if (!el.hasAttribute(name))
        return undefined;
    return el.getAttribute(name) ?? undefined;
}
function getBoolAttr(el, name) {
    const value = el.getAttribute(name);
    if (value === "Yes" || value === "YES")
        return true;
    if (value === "No" || value === "NO")
        return false;
    return undefined;
}
function getIntAttr(el, name) {
    const value = el.getAttribute(name);
    if (value === null || value === undefined)
        return undefined;
    const num = parseInt(value, 10);
    return isNaN(num) ? undefined : num;
}
function getChildElement(parent, tagName) {
    const children = parent.childNodes;
    for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (child.nodeType === 1 && child.tagName === tagName) {
            return child;
        }
    }
    return null;
}
function getChildElements(parent, tagName) {
    const results = [];
    const children = parent.childNodes;
    for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (child.nodeType === 1 && child.tagName === tagName) {
            results.push(child);
        }
    }
    return results;
}
