// @bacons/xcode/build/scheme/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parse = void 0;
/**
 * Parser for Xcode scheme files (.xcscheme)
 *
 * Uses @xmldom/xmldom to parse XML into TypeScript objects.
 */
const xmldom_1 = require("@xmldom/xmldom");
/**
 * Parse an xcscheme XML string into a typed XCScheme object.
 */
function parse(xml) {
    const parser = new xmldom_1.DOMParser();
    const doc = parser.parseFromString(xml, "text/xml");
    const schemeEl = doc.documentElement;
    if (!schemeEl || schemeEl.tagName !== "Scheme") {
        throw new Error("Invalid xcscheme file: root element must be <Scheme>");
    }
    return parseScheme(schemeEl);
}
exports.parse = parse;
function parseScheme(el) {
    const scheme = {};
    scheme.version = getAttr(el, "version");
    scheme.lastUpgradeVersion = getAttr(el, "LastUpgradeVersion");
    const buildActionEl = getChildElement(el, "BuildAction");
    if (buildActionEl) {
        scheme.buildAction = parseBuildAction(buildActionEl);
    }
    const testActionEl = getChildElement(el, "TestAction");
    if (testActionEl) {
        scheme.testAction = parseTestAction(testActionEl);
    }
    const launchActionEl = getChildElement(el, "LaunchAction");
    if (launchActionEl) {
        scheme.launchAction = parseLaunchAction(launchActionEl);
    }
    const profileActionEl = getChildElement(el, "ProfileAction");
    if (profileActionEl) {
        scheme.profileAction = parseProfileAction(profileActionEl);
    }
    const analyzeActionEl = getChildElement(el, "AnalyzeAction");
    if (analyzeActionEl) {
        scheme.analyzeAction = parseAnalyzeAction(analyzeActionEl);
    }
    const archiveActionEl = getChildElement(el, "ArchiveAction");
    if (archiveActionEl) {
        scheme.archiveAction = parseArchiveAction(archiveActionEl);
    }
    return scheme;
}
function parseBuildAction(el) {
    const action = {};
    action.parallelizeBuildables = getBoolAttr(el, "parallelizeBuildables");
    action.buildImplicitDependencies = getBoolAttr(el, "buildImplicitDependencies");
    action.runPostActionsOnFailure = getBoolAttr(el, "runPostActionsOnFailure");
    action.buildArchitectures = getAttr(el, "buildArchitectures");
    const entriesEl = getChildElement(el, "BuildActionEntries");
    if (entriesEl) {
        action.entries = getChildElements(entriesEl, "BuildActionEntry").map(parseBuildActionEntry);
    }
    action.preActions = parseExecutionActions(el, "PreActions");
    action.postActions = parseExecutionActions(el, "PostActions");
    return action;
}
function parseBuildActionEntry(el) {
    const entry = {};
    entry.buildForRunning = getBoolAttr(el, "buildForRunning");
    entry.buildForTesting = getBoolAttr(el, "buildForTesting");
    entry.buildForProfiling = getBoolAttr(el, "buildForProfiling");
    entry.buildForArchiving = getBoolAttr(el, "buildForArchiving");
    entry.buildForAnalyzing = getBoolAttr(el, "buildForAnalyzing");
    const refEl = getChildElement(el, "BuildableReference");
    if (refEl) {
        entry.buildableReference = parseBuildableReference(refEl);
    }
    return entry;
}
function parseBuildableReference(el) {
    return {
        buildableIdentifier: getAttr(el, "BuildableIdentifier") || "primary",
        blueprintIdentifier: getAttr(el, "BlueprintIdentifier") || "",
        buildableName: getAttr(el, "BuildableName") || "",
        blueprintName: getAttr(el, "BlueprintName") || "",
        referencedContainer: getAttr(el, "ReferencedContainer") || "",
    };
}
function parseExecutionActions(parent, containerName) {
    const containerEl = getChildElement(parent, containerName);
    if (!containerEl)
        return undefined;
    const actions = getChildElements(containerEl, "ExecutionAction");
    if (actions.length === 0)
        return undefined;
    return actions.map(parseExecutionAction);
}
function parseExecutionAction(el) {
    const action = {};
    action.actionType = getAttr(el, "ActionType");
    const contentEl = getChildElement(el, "ActionContent");
    if (contentEl) {
        action.actionContent = parseActionContent(contentEl);
    }
    return action;
}
function parseActionContent(el) {
    const content = {};
    content.title = getAttr(el, "title");
    content.scriptText = getAttr(el, "scriptText");
    content.shellToInvoke = getAttr(el, "shellToInvoke");
    const envBuildableEl = getChildElement(el, "EnvironmentBuildable");
    if (envBuildableEl) {
        const refEl = getChildElement(envBuildableEl, "BuildableReference");
        if (refEl) {
            content.environmentBuildable = parseBuildableReference(refEl);
        }
    }
    return content;
}
function parseTestAction(el) {
    const action = {};
    action.buildConfiguration = getAttr(el, "buildConfiguration");
    action.selectedDebuggerIdentifier = getAttr(el, "selectedDebuggerIdentifier");
    action.selectedLauncherIdentifier = getAttr(el, "selectedLauncherIdentifier");
    action.shouldUseLaunchSchemeArgsEnv = getBoolAttr(el, "shouldUseLaunchSchemeArgsEnv");
    action.shouldAutocreateTestPlan = getBoolAttr(el, "shouldAutocreateTestPlan");
    action.preferredScreenCaptureFormat = getAttr(el, "preferredScreenCaptureFormat");
    action.codeCoverageEnabled = getBoolAttr(el, "codeCoverageEnabled");
    action.onlyGenerateCoverageForSpecifiedTargets = getBoolAttr(el, "onlyGenerateCoverageForSpecifiedTargets");
    const testablesEl = getChildElement(el, "Testables");
    if (testablesEl) {
        action.testables = getChildElements(testablesEl, "TestableReference").map(parseTestableReference);
    }
    const testPlansEl = getChildElement(el, "TestPlans");
    if (testPlansEl) {
        action.testPlans = getChildElements(testPlansEl, "TestPlanReference").map(parseTestPlanReference);
    }
    const macroExpansionEl = getChildElement(el, "MacroExpansion");
    if (macroExpansionEl) {
        const refEl = getChildElement(macroExpansionEl, "BuildableReference");
        if (refEl) {
            action.macroExpansion = parseBuildableReference(refEl);
        }
    }
    action.commandLineArguments = parseCommandLineArguments(el);
    action.environmentVariables = parseEnvironmentVariables(el);
    action.preActions = parseExecutionActions(el, "PreActions");
    action.postActions = parseExecutionActions(el, "PostActions");
    return action;
}
function parseTestableReference(el) {
    const ref = {};
    ref.skipped = getBoolAttr(el, "skipped");
    ref.parallelizable = getBoolAttr(el, "parallelizable");
    ref.useTestSelectionWhitelist = getBoolAttr(el, "useTestSelectionWhitelist");
    const buildableRefEl = getChildElement(el, "BuildableReference");
    if (buildableRefEl) {
        ref.buildableReference = parseBuildableReference(buildableRefEl);
    }
    const skippedTestsEl = getChildElement(el, "SkippedTests");
    if (skippedTestsEl) {
        ref.skippedTests = getChildElements(skippedTestsEl, "Test").map((testEl) => ({
            identifier: getAttr(testEl, "Identifier") || "",
        }));
    }
    const selectedTestsEl = getChildElement(el, "SelectedTests");
    if (selectedTestsEl) {
        ref.selectedTests = getChildElements(selectedTestsEl, "Test").map((testEl) => ({
            identifier: getAttr(testEl, "Identifier") || "",
        }));
    }
    return ref;
}
function parseTestPlanReference(el) {
    return {
        reference: getAttr(el, "reference") || "",
        default: getBoolAttr(el, "default"),
    };
}
function parseLaunchAction(el) {
    const action = {};
    action.buildConfiguration = getAttr(el, "buildConfiguration");
    action.selectedDebuggerIdentifier = getAttr(el, "selectedDebuggerIdentifier");
    action.selectedLauncherIdentifier = getAttr(el, "selectedLauncherIdentifier");
    action.launchStyle = getAttr(el, "launchStyle");
    action.useCustomWorkingDirectory = getBoolAttr(el, "useCustomWorkingDirectory");
    action.customWorkingDirectory = getAttr(el, "customWorkingDirectory");
    action.ignoresPersistentStateOnLaunch = getBoolAttr(el, "ignoresPersistentStateOnLaunch");
    action.debugDocumentVersioning = getBoolAttr(el, "debugDocumentVersioning");
    action.debugServiceExtension = getAttr(el, "debugServiceExtension");
    action.allowLocationSimulation = getBoolAttr(el, "allowLocationSimulation");
    action.customLaunchCommand = getAttr(el, "customLaunchCommand");
    action.appClipInvocationURLString = getAttr(el, "appClipInvocationURLString");
    action.askForAppToLaunch = getBoolAttr(el, "askForAppToLaunch");
    action.launchAutomaticallySubstyle = getAttr(el, "launchAutomaticallySubstyle");
    const runnableEl = getChildElement(el, "BuildableProductRunnable");
    if (runnableEl) {
        action.buildableProductRunnable = parseBuildableProductRunnable(runnableEl);
    }
    const remoteRunnableEl = getChildElement(el, "RemoteRunnable");
    if (remoteRunnableEl) {
        action.remoteRunnable = parseRemoteRunnable(remoteRunnableEl);
    }
    const macroExpansionEl = getChildElement(el, "MacroExpansion");
    if (macroExpansionEl) {
        const refEl = getChildElement(macroExpansionEl, "BuildableReference");
        if (refEl) {
            action.macroExpansion = parseBuildableReference(refEl);
        }
    }
    const locationEl = getChildElement(el, "LocationScenarioReference");
    if (locationEl) {
        action.locationScenarioReference = parseLocationScenarioReference(locationEl);
    }
    const storeKitEl = getChildElement(el, "StoreKitConfigurationFileReference");
    if (storeKitEl) {
        action.storeKitConfigurationFileReference =
            parseStoreKitConfigurationFileReference(storeKitEl);
    }
    action.commandLineArguments = parseCommandLineArguments(el);
    action.environmentVariables = parseEnvironmentVariables(el);
    action.preActions = parseExecutionActions(el, "PreActions");
    action.postActions = parseExecutionActions(el, "PostActions");
    return action;
}
function parseBuildableProductRunnable(el) {
    const runnable = {};
    runnable.runnableDebuggingMode = getAttr(el, "runnableDebuggingMode");
    const refEl = getChildElement(el, "BuildableReference");
    if (refEl) {
        runnable.buildableReference = parseBuildableReference(refEl);
    }
    return runnable;
}
function parseRemoteRunnable(el) {
    const runnable = {};
    runnable.runnableDebuggingMode = getAttr(el, "runnableDebuggingMode");
    runnable.bundleIdentifier = getAttr(el, "BundleIdentifier");
    runnable.remotePath = getAttr(el, "RemotePath");
    const refEl = getChildElement(el, "BuildableReference");
    if (refEl) {
        runnable.buildableReference = parseBuildableReference(refEl);
    }
    return runnable;
}
function parseLocationScenarioReference(el) {
    return {
        identifier: getAttr(el, "identifier"),
        referenceType: getAttr(el, "referenceType"),
    };
}
function parseStoreKitConfigurationFileReference(el) {
    return {
        identifier: getAttr(el, "identifier"),
    };
}
function parseProfileAction(el) {
    const action = {};
    action.buildConfiguration = getAttr(el, "buildConfiguration");
    action.shouldUseLaunchSchemeArgsEnv = getBoolAttr(el, "shouldUseLaunchSchemeArgsEnv");
    action.savedToolIdentifier = getAttr(el, "savedToolIdentifier");
    action.useCustomWorkingDirectory = getBoolAttr(el, "useCustomWorkingDirectory");
    action.debugDocumentVersioning = getBoolAttr(el, "debugDocumentVersioning");
    action.askForAppToLaunch = getBoolAttr(el, "askForAppToLaunch");
    action.launchAutomaticallySubstyle = getAttr(el, "launchAutomaticallySubstyle");
    action.appClipInvocationURLString = getAttr(el, "appClipInvocationURLString");
    const runnableEl = getChildElement(el, "BuildableProductRunnable");
    if (runnableEl) {
        action.buildableProductRunnable = parseBuildableProductRunnable(runnableEl);
    }
    action.commandLineArguments = parseCommandLineArguments(el);
    action.environmentVariables = parseEnvironmentVariables(el);
    action.preActions = parseExecutionActions(el, "PreActions");
    action.postActions = parseExecutionActions(el, "PostActions");
    return action;
}
function parseAnalyzeAction(el) {
    const action = {};
    action.buildConfiguration = getAttr(el, "buildConfiguration");
    action.preActions = parseExecutionActions(el, "PreActions");
    action.postActions = parseExecutionActions(el, "PostActions");
    return action;
}
function parseArchiveAction(el) {
    const action = {};
    action.buildConfiguration = getAttr(el, "buildConfiguration");
    action.customArchiveName = getAttr(el, "customArchiveName");
    action.revealArchiveInOrganizer = getBoolAttr(el, "revealArchiveInOrganizer");
    action.preActions = parseExecutionActions(el, "PreActions");
    action.postActions = parseExecutionActions(el, "PostActions");
    return action;
}
function parseCommandLineArguments(el) {
    const argsEl = getChildElement(el, "CommandLineArguments");
    if (!argsEl)
        return undefined;
    const args = getChildElements(argsEl, "CommandLineArgument");
    if (args.length === 0)
        return undefined;
    return args.map((argEl) => ({
        argument: getAttr(argEl, "argument") || "",
        isEnabled: getBoolAttr(argEl, "isEnabled"),
    }));
}
function parseEnvironmentVariables(el) {
    const varsEl = getChildElement(el, "EnvironmentVariables");
    if (!varsEl)
        return undefined;
    const vars = getChildElements(varsEl, "EnvironmentVariable");
    if (vars.length === 0)
        return undefined;
    return vars.map((varEl) => ({
        key: getAttr(varEl, "key") || "",
        value: getAttr(varEl, "value") || "",
        isEnabled: getBoolAttr(varEl, "isEnabled"),
    }));
}
// ============================================================================
// Helper functions
// ============================================================================
function getAttr(el, name) {
    if (!el.hasAttribute(name))
        return undefined;
    return el.getAttribute(name) ?? undefined;
}
function getBoolAttr(el, name) {
    const value = el.getAttribute(name);
    if (value === "YES" || value === "Yes")
        return true;
    if (value === "NO" || value === "No")
        return false;
    return undefined;
}
function getChildElement(parent, tagName) {
    const children = parent.childNodes;
    for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (child.nodeType === 1 && child.tagName === tagName) {
            return child;
        }
    }
    return null;
}
function getChildElements(parent, tagName) {
    const results = [];
    const children = parent.childNodes;
    for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (child.nodeType === 1 && child.tagName === tagName) {
            results.push(child);
        }
    }
    return results;
}
