// @bacons/xcode/build/workspace/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parse = void 0;
/**
 * Parser for Xcode workspace files (contents.xcworkspacedata)
 *
 * Uses @xmldom/xmldom to parse XML into TypeScript objects.
 */
const xmldom_1 = require("@xmldom/xmldom");
/**
 * Parse an xcworkspacedata XML string into a typed XCWorkspace object.
 */
function parse(xml) {
    const parser = new xmldom_1.DOMParser();
    const doc = parser.parseFromString(xml, "text/xml");
    const workspaceEl = doc.documentElement;
    if (!workspaceEl || workspaceEl.tagName !== "Workspace") {
        throw new Error("Invalid xcworkspacedata file: root element must be <Workspace>");
    }
    return parseWorkspace(workspaceEl);
}
exports.parse = parse;
function parseWorkspace(el) {
    const workspace = {};
    workspace.version = getAttr(el, "version");
    const fileRefs = getChildElements(el, "FileRef");
    if (fileRefs.length > 0) {
        workspace.fileRefs = fileRefs.map(parseFileRef);
    }
    const groups = getChildElements(el, "Group");
    if (groups.length > 0) {
        workspace.groups = groups.map(parseGroup);
    }
    return workspace;
}
function parseFileRef(el) {
    return {
        location: getAttr(el, "location") || "",
    };
}
function parseGroup(el) {
    const group = {};
    group.location = getAttr(el, "location");
    group.name = getAttr(el, "name");
    const fileRefs = getChildElements(el, "FileRef");
    if (fileRefs.length > 0) {
        group.fileRefs = fileRefs.map(parseFileRef);
    }
    const nestedGroups = getChildElements(el, "Group");
    if (nestedGroups.length > 0) {
        group.groups = nestedGroups.map(parseGroup);
    }
    return group;
}
// ============================================================================
// Helper functions
// ============================================================================
function getAttr(el, name) {
    if (!el.hasAttribute(name))
        return undefined;
    return el.getAttribute(name) ?? undefined;
}
function getChildElements(parent, tagName) {
    const results = [];
    const children = parent.childNodes;
    for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (child.nodeType === 1 && child.tagName === tagName) {
            results.push(child);
        }
    }
    return results;
}
