// @bacons/xcode/build/xcconfig/index.js
"use strict";
/**
 * Low-level API for parsing and building Xcode build configuration files (.xcconfig).
 *
 * @example
 * ```ts
 * import * as xcconfig from "@bacons/xcode/xcconfig";
 *
 * // Parse xcconfig content
 * const config = xcconfig.parse(`
 *   #include "Base.xcconfig"
 *   PRODUCT_NAME = MyApp
 *   OTHER_LDFLAGS[sdk=iphoneos*] = -framework UIKit
 * `);
 *
 * // Parse file with resolved includes
 * const config = xcconfig.parseFile("./Project.xcconfig");
 *
 * // Flatten build settings (merge includes, apply conditions)
 * const settings = xcconfig.flattenBuildSettings(config, {
 *   sdk: "iphoneos",
 *   arch: "arm64",
 * });
 *
 * // Serialize back to xcconfig format
 * const content = xcconfig.build(config);
 * ```
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.flattenBuildSettings = exports.build = exports.parseFile = exports.parse = void 0;
var parser_1 = require("./parser");
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return parser_1.parse; } });
Object.defineProperty(exports, "parseFile", { enumerable: true, get: function () { return parser_1.parseFile; } });
var writer_1 = require("./writer");
Object.defineProperty(exports, "build", { enumerable: true, get: function () { return writer_1.build; } });
__exportStar(require("./types"), exports);
/**
 * Flatten build settings from an xcconfig, merging includes and applying conditions.
 *
 * Settings are applied in order:
 * 1. Included files (recursively, in order)
 * 2. Settings in the current file
 *
 * Later settings override earlier ones. Settings with conditions are only
 * included if they match the provided sdk/arch/config options.
 *
 * @param config - Parsed XCConfig
 * @param options - Options for filtering and variable resolution
 * @returns Flattened build settings as a key-value map
 */
function flattenBuildSettings(config, options = {}) {
    const settings = {};
    // Process includes first (they are overridden by local settings)
    for (const include of config.includes) {
        if (include.config) {
            const includedSettings = flattenBuildSettings(include.config, options);
            Object.assign(settings, includedSettings);
        }
    }
    // Process local settings
    for (const setting of config.buildSettings) {
        if (matchesConditions(setting.conditions, options)) {
            const value = resolveInherited(setting.value, setting.key, settings);
            settings[setting.key] = value;
        }
    }
    return settings;
}
exports.flattenBuildSettings = flattenBuildSettings;
/**
 * Check if a setting's conditions match the provided options.
 *
 * A conditional setting is only included if:
 * 1. The options specify the relevant filter (sdk, arch, config)
 * 2. The filter value matches the condition's pattern
 *
 * Settings without conditions always match.
 */
function matchesConditions(conditions, options) {
    if (!conditions || conditions.length === 0) {
        return true;
    }
    // All conditions must be satisfied
    return conditions.every((condition) => {
        // If condition specifies sdk, options must also specify sdk and it must match
        if (condition.sdk) {
            if (!options.sdk)
                return false;
            if (!matchWildcard(options.sdk, condition.sdk))
                return false;
        }
        // If condition specifies arch, options must also specify arch and it must match
        if (condition.arch) {
            if (!options.arch)
                return false;
            if (!matchWildcard(options.arch, condition.arch))
                return false;
        }
        // If condition specifies config, options must also specify config and it must match
        if (condition.config) {
            if (!options.config)
                return false;
            if (!matchWildcard(options.config, condition.config))
                return false;
        }
        return true;
    });
}
/**
 * Match a value against a wildcard pattern.
 * Supports * as wildcard for any characters.
 */
function matchWildcard(value, pattern) {
    // Convert wildcard pattern to regex
    const regexPattern = pattern
        .replace(/[.+^${}()|[\]\\]/g, "\\$&") // Escape regex special chars except *
        .replace(/\*/g, ".*"); // Convert * to .*
    const regex = new RegExp(`^${regexPattern}$`, "i");
    return regex.test(value);
}
/**
 * Resolve $(inherited) in a value by prepending the existing value.
 */
function resolveInherited(value, key, existingSettings) {
    const inheritedPattern = /\$\(inherited\)/gi;
    if (!inheritedPattern.test(value)) {
        return value;
    }
    const existingValue = existingSettings[key] ?? "";
    return value.replace(inheritedPattern, existingValue);
}
