// @bacons/xcode/build/xcconfig/parser.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseFile = exports.parse = void 0;
/**
 * Parser for Xcode build configuration files (.xcconfig)
 *
 * Uses regex-based parsing for the simple line-oriented format.
 */
const fs = __importStar(require("node:fs"));
const path = __importStar(require("node:path"));
// Regex patterns for xcconfig parsing
const INCLUDE_REGEX = /^#include(\?)?\s*"(.+)"$/;
const SETTING_REGEX = /^([a-zA-Z_][a-zA-Z0-9_]*(?:\[[^\]]+\])*)\s*=\s*(.*)$/;
const CONDITION_REGEX = /\[([a-zA-Z]+)=([^\]]+)\]/g;
const COMMENT_REGEX = /\/\/.*/;
/**
 * Parse xcconfig content from a string.
 *
 * @param content - The xcconfig file content
 * @returns Parsed XCConfig object (without resolved includes)
 */
function parse(content) {
    const includes = [];
    const buildSettings = [];
    const lines = content.split(/\r?\n/);
    for (const rawLine of lines) {
        // Strip comments (// to end of line)
        const line = rawLine.replace(COMMENT_REGEX, "").trim();
        // Skip empty lines
        if (!line)
            continue;
        // Check for include directive
        const includeMatch = line.match(INCLUDE_REGEX);
        if (includeMatch) {
            const optional = includeMatch[1] === "?";
            const includePath = includeMatch[2];
            includes.push({
                include: { path: includePath, optional },
                resolvedPath: includePath,
                config: undefined,
            });
            continue;
        }
        // Check for setting
        const settingMatch = line.match(SETTING_REGEX);
        if (settingMatch) {
            const keyWithConditions = settingMatch[1];
            const value = settingMatch[2].trim();
            // Extract conditions from key
            const conditions = parseConditions(keyWithConditions);
            const key = keyWithConditions.replace(CONDITION_REGEX, "");
            buildSettings.push({
                key,
                value,
                conditions: conditions.length > 0 ? conditions : undefined,
            });
        }
    }
    return { includes, buildSettings };
}
exports.parse = parse;
/**
 * Parse conditions from a key like "OTHER_LDFLAGS[sdk=iphoneos*][arch=arm64]"
 */
function parseConditions(keyWithConditions) {
    const conditions = [];
    let match;
    const regex = new RegExp(CONDITION_REGEX);
    while ((match = regex.exec(keyWithConditions)) !== null) {
        const conditionType = match[1].toLowerCase();
        const conditionValue = match[2];
        const condition = {};
        if (conditionType === "sdk") {
            condition.sdk = conditionValue;
        }
        else if (conditionType === "arch") {
            condition.arch = conditionValue;
        }
        else if (conditionType === "config") {
            condition.config = conditionValue;
        }
        if (Object.keys(condition).length > 0) {
            conditions.push(condition);
        }
    }
    return conditions;
}
/**
 * Parse an xcconfig file and resolve includes.
 *
 * @param filePath - Path to the xcconfig file
 * @param options - Parse options
 * @returns Parsed XCConfig with resolved includes
 */
function parseFile(filePath, options = {}) {
    const { basePath = path.dirname(filePath), resolveIncludes = true, visitedPaths = new Set(), } = options;
    const absolutePath = path.isAbsolute(filePath)
        ? filePath
        : path.resolve(basePath, filePath);
    // Cycle detection
    if (visitedPaths.has(absolutePath)) {
        throw new Error(`Circular include detected: ${absolutePath}`);
    }
    const content = fs.readFileSync(absolutePath, "utf-8");
    const config = parse(content);
    if (!resolveIncludes) {
        return config;
    }
    // Track this path for cycle detection
    const newVisitedPaths = new Set(visitedPaths);
    newVisitedPaths.add(absolutePath);
    // Resolve includes
    const resolvedIncludes = [];
    const fileDir = path.dirname(absolutePath);
    for (const includeEntry of config.includes) {
        const includePath = includeEntry.include.path;
        const resolvedPath = path.resolve(fileDir, includePath);
        try {
            if (!fs.existsSync(resolvedPath)) {
                if (includeEntry.include.optional) {
                    resolvedIncludes.push({
                        include: includeEntry.include,
                        resolvedPath,
                        config: undefined,
                    });
                    continue;
                }
                throw new Error(`Include file not found: ${resolvedPath}`);
            }
            const includedConfig = parseFile(resolvedPath, {
                basePath: path.dirname(resolvedPath),
                resolveIncludes: true,
                visitedPaths: newVisitedPaths,
            });
            resolvedIncludes.push({
                include: includeEntry.include,
                resolvedPath,
                config: includedConfig,
            });
        }
        catch (error) {
            if (includeEntry.include.optional) {
                resolvedIncludes.push({
                    include: includeEntry.include,
                    resolvedPath,
                    config: undefined,
                });
            }
            else {
                throw error;
            }
        }
    }
    return {
        includes: resolvedIncludes,
        buildSettings: config.buildSettings,
    };
}
exports.parseFile = parseFile;
